nautobot/ipam/models.py

```diff
@@ -966,6 +941,59 @@ def _set_mask_length(self, value):
     mask_length = property(fset=_set_mask_length)
 
 
+class IPAddressToInterface(BaseModel):
+    ip_address = models.ForeignKey("ipam.IPAddress", on_delete=models.CASCADE)
+    interface = models.ForeignKey("dcim.Interface", blank=True, null=True, on_delete=models.CASCADE)
+    vm_interface = models.ForeignKey("virtualization.VMInterface", blank=True, null=True, on_delete=models.CASCADE)
+    source = models.BooleanField(default=False, help_text="Source address")
+    destination = models.BooleanField(default=False, help_text="Destination address")
+    default = models.BooleanField(default=False, help_text="Default address")
+    preferred = models.BooleanField(default=False, help_text="Preferred address")
+    primary = models.BooleanField(default=False, help_text="Primary address")
+    secondary = models.BooleanField(default=False, help_text="Secondary address")
+    standby = models.BooleanField(default=False, help_text="Standby address")
+    primary_for_device = models.BooleanField(default=False, help_text="Primary address for parent device")
+
+    def clean(self):
+        super().clean()
+
+        if self.interface is not None and self.vm_interface is not None:
+            raise ValidationError(
+                {"interface": "Cannot use a single instance to associate to both an Interface and a VMInterface."}
+            )
+
+        if self.interface is None and self.vm_interface is None:
+            raise ValidationError({"interface": "Must associate to either an Interface or a VMInterface."})
+
+        # if primary_for_device=True, set primary_for_device=False for all other instances on the same device in the same address family
+        if self.primary_for_device and self.present_in_database:
+            if self.interface is not None:
+                parent = self.interface.parent
+                parent_q = Q(interface__device=parent)
+            else:
+                parent = self.vm_interface.parent
+                parent_q = Q(vm_interface__virtual_machine=parent)
+
+            if self.ip_address.family == 4 and parent.primary_ip4 != self.ip_address:
+                IPAddressToInterface.objects.filter(
+                    Q(ip_address__in=IPAddress.objects.ip_family(4)) & parent_q
+                ).exclude(id=self.id).update(primary_for_device=False)
+                parent.primary_ip4 = self.ip_address
+                parent.save()
+            if self.ip_address.family == 6 and parent.primary_ip6 != self.ip_address:
+                IPAddressToInterface.objects.filter(
+                    Q(ip_address__in=IPAddress.objects.ip_family(6)) & parent_q
+                ).exclude(id=self.id).update(primary_for_device=False)
+                parent.primary_ip6 = self.ip_address
+                parent.save()
```

Comment:
Two minor optimizations:

1) If the database record was already set as `primary_for_device` (this isn't a change) then there's no need to do this processing and it could be skipped.
2) The `IPAddressToInterface.objects.filter()` could be narrowed down to only select any `primary_for_device=True` instance(s).

Reply:
I rewrote this to optimize the queryset looking for conflicting instances. Regarding item 1, I didn't know if it was possible to check if the field value has changed in the model `clean` method. I thought this was something we only did in serializer `validate` and form `clean` methods.

Reply:
Pattern would be something like:

```python
if self.present_in_database:
    original_data = IPAddressToInterface.objects.get(id=self.id)
    if self.primary_for_device and not original_data.primary_for_device:
        ...
```